File: Android/core/src/main/java/com/khushi/conduit/core/protocol/Envelope.kt
package com.khushi.conduit.core.protocol

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

/**
 * The JSON message carried on channel 0 after the handshake: a command, a
 * response or an event. Spec: Shared/Protocol/README.md §3.
 *
 * Decoding enforces every field rule in the spec, so a malformed envelope is
 * rejected at the edge instead of reaching feature code half-formed. The Mac
 * counterpart is ConduitProtocol/Envelope.swift; both run the same vectors.
 */
object ProtocolVersion {
    const val CURRENT = 1
    const val MINIMUM = 1
}

class ProtocolException(message: String) : Exception(message)

data class ProtocolError(val code: String, val message: String) {
    /** Spec §8: a code this build does not know is handled as `internal`. */
    val normalizedCode: String get() = if (code in ErrorCode.known) code else ErrorCode.INTERNAL
}

sealed interface Outcome {
    data object Success : Outcome
    data class Failure(val error: ProtocolError) : Outcome
}

sealed interface EnvelopeKind {
    data class Command(val requestId: String, val action: String) : EnvelopeKind
    data class Response(val requestId: String, val outcome: Outcome) : EnvelopeKind
    data class Event(val name: String, val epoch: String, val seq: Long, val timestamp: Long?) : EnvelopeKind
}

data class Envelope(
    val kind: EnvelopeKind,
    val payload: JsonObject? = null,
    val version: Int = ProtocolVersion.CURRENT,
) {

    fun toJson(): JsonObject = buildJsonObject {
        put("v", version)
        when (kind) {
            is EnvelopeKind.Command -> {
                put("type", "command")
                put("requestID", kind.requestId)
                put("action", kind.action)
            }
            is EnvelopeKind.Response -> {
                put("type", "response")
                put("requestID", kind.requestId)
                when (val outcome = kind.outcome) {
                    Outcome.Success -> put("status", "success")
                    is Outcome.Failure -> {
                        put("status", "failure")
                        put("error", buildJsonObject {
                            put("code", outcome.error.code)
                            put("message", outcome.error.message)
                        })
                    }
                }
            }
            is EnvelopeKind.Event -> {
                put("type", "event")
                put("event", kind.name)
                put("epoch", kind.epoch)
                put("seq", kind.seq)
                kind.timestamp?.let { put("timestamp", it) }
            }
        }
        payload?.let { put("payload", it) }
    }

    fun encode(): String = toJson().toString()

    companion object {

        fun decode(text: String): Envelope = fromJson(ProtocolJson.parse(text))

        fun fromJson(element: JsonElement): Envelope {
            val json = element as? JsonObject ?: throw ProtocolException("envelope must be a JSON object")

            val version = json.integer("v")?.toInt() ?: throw ProtocolException("v must be an integer")

            val payload = when (val raw = json["payload"]) {
                null -> null
                is JsonObject -> raw
                else -> throw ProtocolException("payload must be a JSON object")
            }

            val kind = when (json.string("type")) {
                "command" -> EnvelopeKind.Command(
                    requestId = json.requireString("requestID"),
                    action = json.requireString("action"),
                )
                "response" -> {
                    val requestId = json.requireString("requestID")
                    val outcome = when (json.string("status")) {
                        "success" -> Outcome.Success
                        "failure" -> {
                            val error = json["error"] as? JsonObject
                                ?: throw ProtocolException("failure needs an error object")
                            Outcome.Failure(ProtocolError(error.requireString("code"), error.requireString("message")))
                        }
                        else -> throw ProtocolException("status must be success or failure")
                    }
                    EnvelopeKind.Response(requestId, outcome)
                }
                "event" -> {
                    val seq = json.integer("seq") ?: throw ProtocolException("seq must be an integer")
                    if (seq < 1) throw ProtocolException("seq starts at 1")
                    EnvelopeKind.Event(
                        name = json.requireString("event"),
                        epoch = json.requireString("epoch"),
                        seq = seq,
                        timestamp = if (json.containsKey("timestamp")) {
                            json.integer("timestamp") ?: throw ProtocolException("timestamp must be an integer")
                        } else null,
                    )
                }
                else -> throw ProtocolException("type must be command, response or event")
            }

            return Envelope(kind = kind, payload = payload, version = version)
        }

        private fun JsonObject.string(key: String): String? =
            (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

        private fun JsonObject.requireString(key: String): String =
            string(key) ?: throw ProtocolException("$key must be a string")

        /** An integer JSON number — not a string, not a fraction. */
        private fun JsonObject.integer(key: String): Long? =
            (this[key] as? JsonPrimitive)?.takeIf { !it.isString }?.longOrNull
    }
}

internal object ProtocolJson {
    private val json = kotlinx.serialization.json.Json { ignoreUnknownKeys = true }

    fun parse(text: String): JsonElement = json.parseToJsonElement(text)
}

File: Shared/ConduitKit/Sources/ConduitProtocol/Envelope.swift
//
//  Envelope.swift
//  ConduitProtocol
//
//  The JSON message carried on channel 0 after the handshake: a command, a
//  response or an event. Spec: Shared/Protocol/README.md §3.
//
//  Decoding enforces every field rule in the spec, so a malformed envelope
//  is rejected at the edge instead of reaching feature code half-formed.
//

import Foundation

public enum ProtocolVersion {
    /// The version this build speaks.
    public static let current = 1
    /// The oldest version this build accepts.
    public static let minimum = 1
}

public struct Envelope: Sendable, Equatable {

    public enum Kind: Sendable, Equatable {
        case command(requestID: String, action: ActionName)
        case response(requestID: String, outcome: Outcome)
        case event(name: EventName, epoch: String, seq: UInt64, timestamp: Int64?)
    }

    public enum Outcome: Sendable, Equatable {
        case success
        case failure(ProtocolError)
    }

    public var version: Int
    public var kind: Kind
    /// Always an object when present (spec §3, `payload`).
    public var payload: [String: JSONValue]?

    public init(version: Int = ProtocolVersion.current, kind: Kind, payload: [String: JSONValue]? = nil) {
        self.version = version
        self.kind = kind
        self.payload = payload
    }

    // MARK: Constructors

    public static func command(_ action: ActionName, payload: [String: JSONValue]? = nil,
                               requestID: String = UUID().uuidString) -> Envelope {
        Envelope(kind: .command(requestID: requestID, action: action), payload: payload)
    }

    public static func success(to requestID: String, payload: [String: JSONValue]? = nil) -> Envelope {
        Envelope(kind: .response(requestID: requestID, outcome: .success), payload: payload)
    }

    public static func failure(to requestID: String, _ error: ProtocolError) -> Envelope {
        Envelope(kind: .response(requestID: requestID, outcome: .failure(error)))
    }

    public static func event(_ name: EventName, epoch: String, seq: UInt64,
                             timestamp: Int64? = nil, payload: [String: JSONValue]? = nil) -> Envelope {
        Envelope(kind: .event(name: name, epoch: epoch, seq: seq, timestamp: timestamp), payload: payload)
    }

    // MARK: Wire

    public func encoded() throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(self)
    }

    public static func decode(_ data: Data) throws -> Envelope {
        try JSONDecoder().decode(Envelope.self, from: data)
    }
}

extension Envelope: Codable {

    private enum CodingKeys: String, CodingKey {
        case v, type, requestID, action, event, status, error, epoch, seq, timestamp, payload
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        version = try c.decode(Int.self, forKey: .v)

        if c.contains(.payload) {
            guard case .object(let fields) = try c.decode(JSONValue.self, forKey: .payload) else {
                throw DecodingError.dataCorruptedError(
                    forKey: .payload, in: c, debugDescription: "payload must be a JSON object")
            }
            payload = fields
        } else {
            payload = nil
        }

        switch try c.decode(String.self, forKey: .type) {
        case "command":
            kind = .command(
                requestID: try c.decode(String.self, forKey: .requestID),
                action: ActionName(try c.decode(String.self, forKey: .action)))

        case "response":
            let requestID = try c.decode(String.self, forKey: .requestID)
            switch try c.decode(String.self, forKey: .status) {
            case "success":
                kind = .response(requestID: requestID, outcome: .success)
            case "failure":
                kind = .response(requestID: requestID,
                                 outcome: .failure(try c.decode(ProtocolError.self, forKey: .error)))
            default:
                throw DecodingError.dataCorruptedError(
                    forKey: .status, in: c, debugDescription: "status must be success or failure")
            }

        case "event":
            let seq = try c.decode(UInt64.self, forKey: .seq)
            guard seq >= 1 else {
                throw DecodingError.dataCorruptedError(
                    forKey: .seq, in: c, debugDescription: "seq starts at 1")
            }
            kind = .event(
                name: EventName(try c.decode(String.self, forKey: .event)),
                epoch: try c.decode(String.self, forKey: .epoch),
                seq: seq,
                timestamp: try c.decodeIfPresent(Int64.self, forKey: .timestamp))

        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: c, debugDescription: "type must be command, response or event")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(version, forKey: .v)

        switch kind {
        case let .command(requestID, action):
            try c.encode("command", forKey: .type)
            try c.encode(requestID, forKey: .requestID)
            try c.encode(action.rawValue, forKey: .action)

        case let .response(requestID, outcome):
            try c.encode("response", forKey: .type)
            try c.encode(requestID, forKey: .requestID)
            switch outcome {
            case .success:
                try c.encode("success", forKey: .status)
            case .failure(let error):
                try c.encode("failure", forKey: .status)
                try c.encode(error, forKey: .error)
            }

        case let .event(name, epoch, seq, timestamp):
            try c.encode("event", forKey: .type)
            try c.encode(name.rawValue, forKey: .event)
            try c.encode(epoch, forKey: .epoch)
            try c.encode(seq, forKey: .seq)
            try c.encodeIfPresent(timestamp, forKey: .timestamp)
        }

        if let payload {
            try c.encode(JSONValue.object(payload), forKey: .payload)
        }
    }
}
